zzz_generated/app-common/app-common/zap-generated/attribute-type.cpp

```diff
@@ -0,0 +1,185 @@
+        return 8;
+    case ZCL_ENUM8_ATTRIBUTE_TYPE:
+        return 1;
+    case ZCL_ENUM16_ATTRIBUTE_TYPE:
+        return 2;
+    case ZCL_PRIORITY_ATTRIBUTE_TYPE:
+        return 1;
+    case ZCL_STATUS_ATTRIBUTE_TYPE:
+        return 1;
+    case ZCL_SINGLE_ATTRIBUTE_TYPE:
+        return 4;
+    case ZCL_DOUBLE_ATTRIBUTE_TYPE:
+        return 8;
+    case ZCL_OCTET_STRING_ATTRIBUTE_TYPE:
+        return 0;
+    case ZCL_CHAR_STRING_ATTRIBUTE_TYPE:
+        return 0;
+    case ZCL_LONG_OCTET_STRING_ATTRIBUTE_TYPE:
+        return 0;
+    case ZCL_LONG_CHAR_STRING_ATTRIBUTE_TYPE:
+        return 0;
+    case ZCL_ARRAY_ATTRIBUTE_TYPE:
+        return 0;
+    case ZCL_STRUCT_ATTRIBUTE_TYPE:
+        return 0;
+    case ZCL_GROUP_ID_ATTRIBUTE_TYPE:
+        return 2;
+    case ZCL_ENDPOINT_NO_ATTRIBUTE_TYPE:
+        return 2;
+    case ZCL_VENDOR_ID_ATTRIBUTE_TYPE:
+        return 2;
+    case ZCL_DEVTYPE_ID_ATTRIBUTE_TYPE:
+        return 4;
+    case ZCL_FABRIC_ID_ATTRIBUTE_TYPE:
+        return 8;
+    case ZCL_FABRIC_IDX_ATTRIBUTE_TYPE:
+        return 1;
+    case ZCL_ENTRY_IDX_ATTRIBUTE_TYPE:
+        return 2;
+    case ZCL_DATA_VER_ATTRIBUTE_TYPE:
+        return 4;
+    case ZCL_EVENT_NO_ATTRIBUTE_TYPE:
+        return 8;
+    case ZCL_SEMTAG_ATTRIBUTE_TYPE:
+        return 4;
+    case ZCL_NAMESPACE_ATTRIBUTE_TYPE:
+        return 1;
+    case ZCL_TAG_ATTRIBUTE_TYPE:
+        return 1;
+    case ZCL_SYSTIME_US_ATTRIBUTE_TYPE:
+        return 8;
+    case ZCL_SYSTIME_MS_ATTRIBUTE_TYPE:
+        return 8;
+    case ZCL_ELAPSED_S_ATTRIBUTE_TYPE:
+        return 4;
+    case ZCL_TEMPERATURE_ATTRIBUTE_TYPE:
+        return 2;
+    case ZCL_POWER_MW_ATTRIBUTE_TYPE:
+        return 8;
+    case ZCL_AMPERAGE_MA_ATTRIBUTE_TYPE:
+        return 8;
+    case ZCL_VOLTAGE_MV_ATTRIBUTE_TYPE:
+        return 8;
+    case ZCL_ENERGY_MWH_ATTRIBUTE_TYPE:
+        return 8;
+    case ZCL_TOD_ATTRIBUTE_TYPE:
+        return 4;
+    case ZCL_DATE_ATTRIBUTE_TYPE:
+        return 4;
+    case ZCL_EPOCH_US_ATTRIBUTE_TYPE:
+        return 8;
+    case ZCL_EPOCH_S_ATTRIBUTE_TYPE:
+        return 4;
+    case ZCL_POSIX_MS_ATTRIBUTE_TYPE:
+        return 8;
+    case ZCL_PERCENT_ATTRIBUTE_TYPE:
+        return 1;
+    case ZCL_PERCENT100THS_ATTRIBUTE_TYPE:
+        return 2;
+    case ZCL_CLUSTER_ID_ATTRIBUTE_TYPE:
+        return 4;
+    case ZCL_ATTRIB_ID_ATTRIBUTE_TYPE:
+        return 4;
+    case ZCL_FIELD_ID_ATTRIBUTE_TYPE:
+        return 4;
+    case ZCL_EVENT_ID_ATTRIBUTE_TYPE:
+        return 4;
+    case ZCL_COMMAND_ID_ATTRIBUTE_TYPE:
+        return 4;
+    case ZCL_ACTION_ID_ATTRIBUTE_TYPE:
+        return 1;
+    case ZCL_TRANS_ID_ATTRIBUTE_TYPE:
+        return 4;
+    case ZCL_NODE_ID_ATTRIBUTE_TYPE:
+        return 8;
+    case ZCL_IPADR_ATTRIBUTE_TYPE:
+        return 0;
+    case ZCL_IPV4ADR_ATTRIBUTE_TYPE:
+        return 4;
+    case ZCL_IPV6ADR_ATTRIBUTE_TYPE:
+        return 16;
+    case ZCL_IPV6PRE_ATTRIBUTE_TYPE:
```

Comment:
This doesn't seem right

Reply:
In the XML this is a "composite" type without a fixed size set.  In the spec it's sized to "1 to 17 bytes".  So this is correct: we have no idea what the size of this thing will be, just like random char/octet strings....

Reply:
So how does this work in the code? How can you generate the right code? Does it take any value whatsoever? How is the size used? The write attribute call to Ember does not take a size...

Reply:
For things that return 0 here we can't optimize out "no-op" writes because we have no way to figure out what the "old value" really is....  That's why the code that does the no-op optimization is skipped if the size is 0.